packages/core/src/tools/instrumentMethod.ts

```diff
@@ -0,0 +1,25 @@
+export function instrumentMethod<S extends { [key: string]: any }, N extends keyof S>(
+  source: S,
+  name: N,
+  replacementFactory: (original: S[N]) => (this: S, ...args: Parameters<S[N]>) => ReturnType<S[N]>
+) {
+  const original = source[name]
+
+  let replacement = replacementFactory(original)
+
+  const replacementWrapper = function (this: S): ReturnType<S[N]> {
```

Comment:
Good question! A replacement method would be less convenient, because we'd like to pass the original method somehow. So we could pass the 'original' as first argument of the replacement, like: 

```ts
instrumentMethod(window, 'fetch', (original, ...fetchArguments) => {
  original(...fetchArguments)
})
```

But this has a slightly bigger performance impact since we need to build the `fetchArguments` array, and I'd like to keep the ability to pass `arguments` directly, like: `original.apply(this, arguments)`.

Reply:
It's been a long time since I didn't run a microbenchmark, but here you go:
![Screenshot 2021-10-07 at 12 00 36](https://user-images.githubusercontent.com/61787/136363422-4febb5d1-41c1-4bf9-b18f-0f0e4d20e079.png)

Using spread instead of `arguments` is 98% slower in this small test case. See [the original TS source](https://www.typescriptlang.org/play?target=1#code/GYVwdgxgLglg9mABMOcAUBDANIgRgSkQG8BYAKEUQCcBTKEKpDRAaj3IF9zzRJYFEABwxQIACwDKg2hgAmaOFRgBzGGAwAbHLUEaMEGgFsaYKIVIVqdBkl7R4SNOfKVKteoyu79Rk1AB0EJoaaFBiMADOOIoqapo4-okYVMogxqYR+C6IXGS55BAIEVBeegbpUFIysogAvEIi4lU0cmgocDh2-I4xqupaiIn+ycqZxNnuNoi9cRrDgroAnqHhUYgjmZxZZNxkXQ4NomIAgilpfhEKSn3xpT4VAGL6UIqLzpaTnjplvqZP0K8rrF+tt8jsyIUwMU7uU-KdUhUInVDuJ4ecMm1UDggTcNIRagA+ZDgewCJzjD7WTwzfrzJYrSI4EboqCbPLbXbfe5ws6ItAARhwACYOWQubDTM1WoLECKgA) to have a bette understanding on the test I run.